assign struct with out of order type declaration

The following program `sample.go` triggers a panic:

```go
package main

type T1 struct {
    A []T2
    M map[uint64]T2
}

type T2 struct {
    name string
}   
    
var t = T1{}
    
func main() {
    println("ok")
}
```

Expected result:
```console
$ go run ./sample.go
ok
```

Got:
```console
$ yaegi ./sample.go
./sample.go:12:9: panic
panic: reflect.Set: value of type struct { A []interface {}; M map[uint64]interface {} } is not assignable to type struct { A []struct { Xname string }; M map[uint64]interface {} } [recovered]
        panic: reflect.Set: value of type struct { A []interface {}; M map[uint64]interface {} } is not assignable to type struct { A []struct { Xname string }; M map[uint64]interface {} }

goroutine 1 [running]:
github.com/containous/yaegi/interp.runCfg.func1(0xc00026c000, 0xc00018dc00)
        /home/marc/go/src/github.com/containous/yaegi/interp/run.go:112 +0x218
panic(0xc9ba60, 0xc0000767c0)
        /usr/lib/go/src/runtime/panic.go:679 +0x1b2
reflect.Value.assignTo(0xc00008a4e0, 0xc00000e960, 0x199, 0xdda212, 0xb, 0xc00008a420, 0x0, 0xc00003e780, 0x0, 0xc00000e960)
        /usr/lib/go/src/reflect/value.go:2403 +0x432
reflect.Value.Set(0xc00008a420, 0xc00000e920, 0x199, 0xc00008a4e0, 0xc00000e960, 0x199)
        /usr/lib/go/src/reflect/value.go:1537 +0xbd
github.com/containous/yaegi/interp.compositeLit.func1(0xc00026c000, 0x1543040)
        /home/marc/go/src/github.com/containous/yaegi/interp/run.go:1579 +0x3e5
github.com/containous/yaegi/interp.runCfg(0xc00018dc00, 0xc00026c000)
        /home/marc/go/src/github.com/containous/yaegi/interp/run.go:118 +0x7a
github.com/containous/yaegi/interp.(*Interpreter).run(0xc0002629a0, 0xc00018c100, 0x0)
        /home/marc/go/src/github.com/containous/yaegi/interp/run.go:96 +0x239
github.com/containous/yaegi/interp.(*Interpreter).Eval(0xc0002629a0, 0xc000272000, 0x89, 0x289, 0xc000272000, 0x89, 0x0, 0x1)
        /home/marc/go/src/github.com/containous/yaegi/interp/interp.go:364 +0x3ef
main.main()
        /home/marc/go/src/github.com/containous/yaegi/cmd/yaegi/yaegi.go:82 +0x42e
```

--- a/_test/struct29.go
+++ b/_test/struct29.go
@@ -1,0 +1,19 @@
+package main
+
+type T1 struct {
+	A []T2
+	B []T2
+}
+
+type T2 struct {
+	name string
+}
+
+var t = T1{}
+
+func main() {
+	println("ok")
+}
+
+// Output:
+// ok

--- a/_test/struct30.go
+++ b/_test/struct30.go
@@ -1,0 +1,19 @@
+package main
+
+type T1 struct {
+	A []T2
+	M map[uint64]T2
+}
+
+type T2 struct {
+	name string
+}
+
+var t = T1{}
+
+func main() {
+	println("ok")
+}
+
+// Output:
+// ok

--- a/interp/gta.go
+++ b/interp/gta.go
@@ -30,7 +30,7 @@ func (interp *Interpreter) gta(root *node, rpath string) ([]*node, error) {
 		case defineStmt:
 			var atyp *itype
 			if n.nleft+n.nright < len(n.child) {
-				// Type is declared explicetly in the assign expression.
+				// Type is declared explicitly in the assign expression.
 				if atyp, err = nodeType(interp, sc, n.child[n.nleft]); err != nil {
 					return false
 				}

--- a/interp/gta.go
+++ b/interp/gta.go
@@ -30,6 +30,7 @@ func (interp *Interpreter) gta(root *node, rpath string) ([]*node, error) {
 		case defineStmt:
 			var atyp *itype
 			if n.nleft+n.nright < len(n.child) {
+				// Type is declared explicetly in the assign expression.
 				if atyp, err = nodeType(interp, sc, n.child[n.nleft]); err != nil {
 					return false
 				}

--- a/interp/run.go
+++ b/interp/run.go
@@ -239,7 +239,7 @@ func isRecursiveStruct(t *itype, rtype reflect.Type) bool {
 	if t.cat == structT && rtype.Kind() == reflect.Interface {
 		return true
 	}
-	if t.cat == ptrT {
+	if t.cat == ptrT && t.rtype != nil {
 		return isRecursiveStruct(t.val, t.rtype.Elem())
 	}
 	return false

--- a/interp/type.go
+++ b/interp/type.go
@@ -885,7 +885,9 @@ func (t *itype) refType(defined map[string]bool) reflect.Type {
 			panic(err)
 		}
 	}
-	if t.val != nil && defined[t.val.name] {
+	if t.val != nil && defined[t.val.name] && !t.val.incomplete && t.val.rtype == nil {
+		// Replace reference to self (direct or indirect) by an interface{} to handle
+		// recursive types with reflect.
 		t.val.rtype = interf
 	}
 	switch t.cat {

